### PR TITLE
Add support for foreign members to FeatureWriter.

### PR DESCRIPTION
**Pull request by dabreegster**
- [X] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/master/CODE_OF_CONDUCT.md).
- [X] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.

### CODE REVIEW

**Comment by dabreegster** [comment]
Successfully using this in https://github.com/dabreegster/routing-engines/blob/c58cc43832714c3c2ca17105f184b2885313eddd/aggregate_routes/src/osm2network.rs#L350

**Comment by michaelkirk** [comment]
Hrmm... my inline comment got lost in the ether.

What would you think about changing the API slightly to avoid building up the intermediate JsonObject just to pass it in?

```
pub fn write_foreign_member(&mut self, key: &str, foreign_member: &JsonValue) -> Result<()>;
```

File: src/feature_writer.rs [hunk 1]
```diff
@@ -8,6 +8,7 @@ use std::io::Write;
 enum State {
     New,
     Started,
```
Comment (michaelkirk) [inline src/feature_writer.rs:10]:
```suggestion
    WritingFeatures,
```

File: src/feature_writer.rs [hunk 2]
```diff
@@ -8,6 +8,7 @@ use std::io::Write;
...
+    StartedWithForeignMembers,
```
Comment (michaelkirk) [inline src/feature_writer.rs:11]:
```suggestion
    WritingForeignMembers,
```

Maybe clearer?

File: src/feature_writer.rs (hunk 1 above)
Comment (dabreegster) [reply src/feature_writer.rs]: Fixed. Thanks, much more clear

**Comment by dabreegster** [comment]
> What would you think about changing the API slightly to avoid building up the intermediate JsonObject just to pass it in?

Good idea, done. It was an odd API before to only allow one of the calls. Writing one key at a time is much nicer.

File: src/feature_writer.rs [hunk 3]
```diff
@@ -158,15 +166,51 @@ impl<W: Write> FeatureWriter<W> {
...
+                self.write_str(r#" "features": ["#)?;
+                self.state = State::WritingFeatures;
+            }
         }
         to_feature_writer(&mut self.writer, value)
     }
 
+    /// Write a [foreign member](https://datatracker.ietf.org/doc/html/rfc7946#section-6) to the
+    /// output stream. This must be done before appending any features.
+    pub fn write_foreign_member(&mut self, key: &str, value: &JsonValue) -> Result<()> {
```
Comment (michaelkirk) [inline src/feature_writer.rs:184]:
```suggestion
    pub fn write_foreign_member<T: ?Sized + Serialize>(&mut self, key: &str, value: &T) -> Result<()> {
```

File: src/feature_writer.rs [hunk 4]
```diff
@@ -378,6 +422,62 @@ mod tests {
...
     }
 
+    #[test]
+    fn write_foreign_members() {
+        let mut buffer: Vec<u8> = vec![];
+        {
+            let mut writer = FeatureWriter::from_writer(&mut buffer);
+
+            writer
+                .write_foreign_member("extra", &json!("string"))
```
Comment (michaelkirk) [inline src/feature_writer.rs:432]:
```suggestion
                .write_foreign_member("extra", "string")
```

File: src/feature_writer.rs [hunk 5]
```diff
@@ -378,6 +422,62 @@ mod tests {
...
+                .unwrap();
+            writer
+                .write_foreign_member("list", &json!([1, 2, 3]))
```
Comment (michaelkirk) [inline src/feature_writer.rs:435]:
```suggestion
                .write_foreign_member("list", &[1, 2, 3])
```

**Review by michaelkirk — CHANGES_REQUESTED** [review]
This gets a little nicer to use if we allow the same value types as `serde_json::to_writer`.

**Comment by dabreegster** [comment]
> This gets a little nicer to use if we allow the same value types as serde_json::to_writer.

Good point, thanks! Done

**Review by michaelkirk — APPROVED** [review]
Thank you!

**Pull request merged**